test/test_type_promotion.py

```diff
@@ -0,0 +1,155 @@
+import torch
+import unittest
+
+from common_utils import TestCase, run_tests, load_tests
+
+# load_tests from common_utils is used to automatically filter tests for
+# sharding on sandcastle. This line silences flake warnings
+load_tests = load_tests
+
+class TestTypePromotion(TestCase):
+
+    def setUp(self):
+        super(TestTypePromotion, self).setUp()
+        torch.set_default_dtype(torch.float32)
+        self.device = 'cpu'
+
+    # In-place operations don't promote.
+    # int+float==float but int.add_(float)==int.
+    # Promoting inplace would require re-allocating and copying the memory of the
+    # tensor data, since element size could change.
+    def test_inplace(self):
+        int_tensor = torch.ones([4, 4, 4], dtype=torch.int32, device=self.device)
+
+        self.assertRaisesRegex(RuntimeError, "can't be cast to", lambda: int_tensor.add_(1.5))
+
+        expected = torch.ones([4, 4, 4], device=self.device)
+        expected = expected.to(torch.int32)
+
+        long_tensor = torch.ones([4, 4, 4], dtype=torch.int64, device=self.device)
+        int_tensor.add_(long_tensor)
+        int_tensor.add_(1)
+        self.assertEqual(int_tensor, expected + 2)
```

Comment:
does this verify that the dtype is still the same?  I seem to remember assertEqual doesn't actually do that, and seems worth checking in this case.

Reply:
I can add a check to verify the dtype didn't change, though I'm not sure if it's meaningful here since this test is for the in-place operation. I don't think it's possible to in-place modify the dtype of a tensor.

Reply:
right, but that's something that could potentially be screwed up.

Reply:
added the check